### PR TITLE
fix(optimizer-geometry): use exact power-of-two rescaling in spectral_matrix_sign_transaction (#2379)

### LINKED ISSUE #2379: bug(optimizer-geometry): the 1e-12 Frobenius floor launders small matrices into a numerically zero matrix sign
## Summary

`spectral_matrix_sign` divides by a fixed `1e-12` magnitude floor instead of the matrix's own Frobenius norm, so any matrix whose norm falls below that floor is **laundered into a numerically zero "valid" sign** rather than orthogonalized. The function's docstring promises the opposite:

> Frobenius normalization places every singular value in its convergence interval and preserves an exact zero for a zero matrix.

`alberta_framework/evaluation/optimizer_geometry.py:174-176`:

```python
norm = jnp.linalg.norm(value)
valid = jnp.all(jnp.isfinite(value)) & jnp.isfinite(norm)
x = value / jnp.maximum(norm, jnp.asarray(1e-12, dtype=value.dtype))
```

When `norm < 1e-12` the divisor is the floor, not the norm, so `‖x‖_F = ‖M‖_F / 1e-12 ≪ 1`. The NS5 recurrence `f(X) = 3/2 X − 1/2 XX^T X` is then applied to singular values far below its convergence interval: each step multiplies them by ≈1.5, so five steps return ≈`7.59 · σ / 1e-12` — a matrix that is numerically zero — while the transaction still reports `valid=True`.

In float32 the floor is reached earlier than the exponent range suggests, because the sum of squares underflows: for entries near `1e-20`, `jnp.linalg.norm` returns exactly `0.0` for a plainly nonzero matrix.

## Reproduction

```python
import jax
import jax.numpy as jnp
from alberta_framework.evaluation.optimizer_geometry import spectral_matrix_sign

m = jnp.array([[2.0, 1.0], [0.5, -1.0]], dtype=jnp.float32)
for scale in (1.0, 1e-13, 1e-15, 1e-20, 1e-25):
    tiny = m * jnp.asarray(scale, dtype=jnp.float32)
    print(scale, float(jnp.linalg.norm(tiny)), float(jnp.linalg.norm(spectral_matrix_sign(tiny))))
```

```
scale     norm(input)      norm(returned sign)   expected
1e+00     2.500000e+00     1.414202e+00          1.414202e+00
1e-13     2.500000e-13     1.201002e+00          1.414202e+00
1e-15     2.500000e-15     1.898316e-02          1.414202e+00
1e-20     0.000000e+00     1.898437e-07          1.414202e+00
1e-25     0.000000e+00     1.898438e-12          1.414202e+00
```

Every row is the same matrix up to a positive scalar, so every row must return the same sign. `valid` is `True` in all five.

Singular values of the returned "sign" (computed in float64):

| input scale | singular values | orthogonal? |
| --- | --- | --- |
| `1.0` | `[0.99999999, 0.99998346]` | yes |
| `1e-13` | `[0.96671583, 0.71264756]` | no |
| `1e-25` | `[1.69801409e-12, 8.49007043e-13]` | no — numerically zero |

The reverse-mode derivative is also lost: `jax.jacrev(spectral_matrix_sign)` is non-finite at `1e-25` scale and finite at unit scale.

## Consumer impact

`muon_ogd_dual_update_transaction` uses this sign as the update direction (`optimizer_geometry.py:260` and `:274`). On the documented empty-constraint path — pinned by `test_muon_ogd_empty_constraints_reduces_exactly_to_ns5` as reducing exactly to NS5 — the update vanishes with gradient scale instead of staying scale-free:

| gradient scale | `‖update‖_F` |
| --- | --- |
| `1.0` | `1.414202e+00` |
| `1e-13` | `1.201002e+00` |
| `1e-25` | `1.898438e-12` |

A scale-free orthogonalized direction is the entire point of the Muon-OGD step, so a small-magnitude momentum matrix silently produces no step rather than a unit-scale one.

The frozen streaming benchmark does not expose this: `_frozen_stream()` matrices have Frobenius norms between `1.15` and `2.59`, all far above the floor, which is why every published arm metric is unaffected.

## Why the floor is the wrong instrument

The existing `test_geometry_float32_overflow_is_invalid_not_laundered` pins the policy for the other end of the exponent range: when the input's own arithmetic overflows, the transaction must report `valid=False` rather than launder a plausible-looking result. The small end does the opposite — it returns a fabricated near-zero matrix with `valid=True`. A magnitude floor cannot distinguish "genuinely zero" from "small but well-conditioned"; only the norm can, and the norm is unusable in the input dtype at that magnitude.

## Proposed fix

Rescale by an exact power of two before normalizing, which is the pattern already used for the same hazard in `alberta_framework/utils/nexting.py`, `alberta_framework/core/oak.py`, `alberta_framework/core/state_builder.py`, and `alberta_framework/benchmarks/ipmnist_screening.py`:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(value)))
rescaled = jnp.ldexp(value, -exponent)
rescaled_norm = jnp.linalg.norm(rescaled)
positive = rescaled_norm > 0
x = jnp.where(
    positive,
    rescaled / jnp.where(positive, rescaled_norm, jnp.ones_like(rescaled_norm)),
    jnp.zeros_like(rescaled),
)
```

The largest entry lands in `[0.5, 1)`, so the sum of squares can neither overflow nor underflow, and an exact zero is preserved by the zero test rather than by a magnitude floor. Because power-of-two scaling is exact, the quotient is bit-identical to `value / norm` for every input whose norm was already representable — verified bit-identical on 200 random float32 matrices spanning `2**-30` to `2**30`. `norm` stays in place as the overflow signal so the pinned invalid-on-overflow behaviour is unchanged.

## Environment

JAX 0.11.0, jaxlib 0.11.0, numpy 2.5.1, CPython 3.13.14, Windows 11 AMD64, `jax_enable_x64` False (default).

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2379.

## Summary
In `spectral_matrix_sign_transaction` (`alberta_framework/evaluation/optimizer_geometry.py`), normalization divided by `jnp.maximum(norm, 1e-12)`. When a matrix's Frobenius norm fell below `1e-12` (or entries were near `1e-20` where standard float32 squared sums underflow), the divisor became the `1e-12` floor instead of the matrix's norm. This caused the NS5 recurrence to run far below its convergence interval, returning a numerically zero matrix with `valid=True`.

## Proposed Changes
1. Used exact power-of-two rescaling (`jnp.frexp` and `jnp.ldexp`) before computing the Frobenius norm in `spectral_matrix_sign_transaction`.
2. Normalized the rescaled matrix whenever `rescaled_norm > 0.0`, preserving scale-invariance across the entire float32 dynamic range while preserving exact zero on zero matrices.
3. Added unit tests in `tests/test_optimizer_geometry.py` verifying that `spectral_matrix_sign_transaction` produces scale-invariant unit-scale orthogonal signs across inputs scaled down to `1e-25`.

## Test Plan
- `uv run pytest tests/test_optimizer_geometry.py` (33 passed)
- `uv run ruff check alberta_framework/evaluation/optimizer_geometry.py tests/test_optimizer_geometry.py` (clean)
- `uv run mypy alberta_framework/evaluation/optimizer_geometry.py` (clean)

### CODE REVIEW

**Review by OutstandingRetard — APPROVED** [review]
Scoped review of #2785 (fix(optimizer-geometry): use exact power-of-two rescaling in spectral_matrix_sign_transaction (#2379)).
Touches 2 file(s): `alberta_framework/evaluation/optimizer_geometry.py`, `tests/test_optimizer_geometry.py`. Lead change `alberta_framework/evaluation/optimizer_geometry.py` (+10/-1).
Matches the stated fix and is ready for maintainer merge.
No blocking concerns.

**Review by agent-cortex — APPROVED** [review]
## Approve — verified by direct reproduction of the real head code

I independently reproduced the defect and the fix by loading the **actual** `alberta_framework/evaluation/optimizer_geometry.py` from both `baseSha c9aba7b` and `headSha 2627ee3` directly (the module is self-contained: stdlib + jax/numpy only, no relative imports), on jax/jaxlib 0.7.1.

### Scope matches issue #2379
The change replaces the `1e-12` magnitude floor (`x = value / jnp.maximum(norm, 1e-12)`) with exact power-of-two rescaling via `jnp.frexp`/`jnp.ldexp` before Frobenius normalization. That is precisely the "laundered into a numerically zero valid sign" bug the issue describes, and nothing beyond it.

### Defect reproduced on base, fix confirmed on head
Same matrix `[[2,1],[0.5,-1]]` scaled by a positive scalar — every row must return the same sign:

```
==== BASE (old 1e-12 floor) ====
   scale      in_norm    sign_norm  valid  singular_values
   1e+00   2.5000e+00   1.4142e+00   True  [1.       0.999983]
   1e-15   2.5000e-15   1.8983e-02   True  [0.016979 0.00849 ]   <- not orthogonal
   1e-20   0.0000e+00   1.8984e-07   True  [0. 0.]               <- numerically zero, valid=True
   1e-25   0.0000e+00   1.8984e-12   True  [0. 0.]
==== HEAD (frexp/ldexp) ====
   1e+00   2.5000e+00   1.4142e+00   True  [1.       0.999983]
   1e-15   2.5000e-15   1.4142e+00   True  [1.       0.999983]
   1e-20   0.0000e+00   1.4142e+00   True  [1.       0.999983]   <- scale invariant
   1e-25   0.0000e+00   1.4142e+00   True  [1.       0.999983]
zero-matrix -> norm 0.0 valid True   (both base and head: exact zero preserved)
```

### The added regression test is genuine (fails pre-fix, passes post-fix)
Replaying the exact assertions of `test_spectral_matrix_sign_is_scale_invariant_for_small_matrices` (tol `1e-3`) against the real modules:

```
head: scale=1e-15 valid=True |diff|=0.00e+00 pass=True  ... ALL PASS: True
base: scale=1e-15 valid=True |diff|=1.40e+00 pass=False
base: scale=1e-20 valid=True |diff|=1.41e+00 pass=False
```

So the test would have caught the bug; it is not a tautology.

### Probes beyond the diff (all clean)
- Non-finite input `[[inf,1],[2,3]]` → `valid=False`, sign 0 (unchanged).
- Non-square `3x2` tiny (`1e-18`) matrix → correct transpose path, valid orthogonal sign.
- Zero matrix → exact zero preserved with `valid=True`.
- The nested `jnp.where(positive, .../ jnp.where(positive, rescaled_norm, ones), zeros)` is the standard NaN-safe double-where; correct.

### One non-blocking observation (out of scope, NOT a regression)
For this input, matrices scaled `>= ~1e19` return `valid=False, norm=0` on **both** base and head — identical behavior. Cause: the retained validity line still computes `norm = jnp.linalg.norm(value)` on the **un-rescaled** matrix, whose float32 squared-Frobenius sum overflows to `inf` (`isfinite(norm)` → False). I confirmed the NS5 recurrence itself converges fine at `1e20` (x_norm → 1.4142) when fed the rescaled matrix, so the large-magnitude rejection is purely the un-rescaled validity norm. This is pre-existing, matches base, and is orthogonal to #2379 (small-norm floor). A future follow-up could compute the validity/finiteness check on the rescaled matrix too, but it is not required here.

### Execution-proof caveat
The full `pytest tests/test_optimizer_geometry.py` suite could not be executed in my sandbox: `alberta_framework`'s import chain uses PEP 695 syntax (`def f[T]`, `type` aliases) and the image only has Python 3.11 (repo `requires-python >= 3.12`). I therefore executed the real head module directly and replayed the added test's exact assertions rather than the pytest harness. The touched module itself is PEP 695-free, so this is faithful to the shipped code.

Correctness, scope, and the regression guard all check out; no blocking defects found.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@bc84c0527374221f5c9f40ede4fbe2e62f703d8c:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Comment by lalalune** [comment]
Superseded by merged #2380. Reviewed this matrix-sign rescaling diff: exact power-of-two normalization now replaces the fixed 1e-12 floor. The selected implementation has 17 passing regressions covering scale invariance, JIT, gradients, float16/bfloat16, zero input, and the Muon-OGD consumer, with required CI green. Keeping another equivalent normalization branch would add no useful behavior.
